firestore/firestore_snippets/listen_document.go

```diff
@@ -19,33 +19,42 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"time"
 
 	"cloud.google.com/go/firestore"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
-// listenDocument listens to a single document, returning data
-// after the first snapshot.
+// listenDocument listens to a single document.
 func listenDocument(w io.Writer, projectID string) error {
 	// projectID := "project-id"
-	ctx := context.Background()
+	// Сontext with timeout stops listening to changes.
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
 
 	client, err := firestore.NewClient(ctx, projectID)
 	if err != nil {
 		return fmt.Errorf("firestore.NewClient: %v", err)
 	}
 	defer client.Close()
 
-	dsnap := client.Collection("cities").Doc("SF").Snapshots(ctx)
-
-	snap, err := dsnap.Next()
-	if !snap.Exists() {
-		return fmt.Errorf("current data: null")
-	}
-	if err != nil {
-		return fmt.Errorf("listen failed: %v", err)
+	it := client.Collection("cities").Doc("SF").Snapshots(ctx)
+	for {
+		snap, err := it.Next()
+		// DeadlineExceeded will be returned when ctx is cancelled.
+		if status.Code(err) == codes.DeadlineExceeded {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("Snapshots.Next: %v", err)
+		}
+		if !snap.Exists() {
+			fmt.Fprintf(w, "Document no longer exists\n")
+			return nil
+		}
+		fmt.Fprintf(w, "Received document snapshot: %v\n", snap.Data())
```

Comment:
Should there be a sleep between loop iterations? @tbpg thoughts?

Reply:
I don't think so. Will stop when the context is cancelled.